### ISSUE TITLE
Add abiliy to customize commit message format

### ISSUE THREAD

**Issue by asfaltboy** [allenta/gitlab-listener#6]
For existing projects which already have a deeply rooted commit message format ideology, it would be very helpful to be able to change the format. Perhaps adding an option for supplying a custom commit message with <placeholder> values for the required/optional parameters so that the required values can be extracted.

**Comment by carlosabalde** [comment]
Hi,

It would be helpful to have some examples of other possible formats. I guess the main difference between formats would be the way of referencing issues (i.e. lists of ids at the begging of the message vs. lists of issues referenced using some modifier like `#refs`). Anyway, any other ideas are welcome.

**Comment by asfaltboy** [comment]
Actually you are right, for easier user facing interface, it would be simpler to display common options for commit formats.

Myself, I like to follow [the commit format angular.js uses](https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format), where the reference to an issue being closed is placed in the `<footer>` (using a github like reference <Close PROJ-123>). Ideally an implementation would allow both reference and action using predefined keywords  and user defined mapping: (Close PROJ-123 => Resolve Issue PROJ-123). This requires that the plugin options would contain a list of keyword => action mapping formats.

In any case, IMHO such "meta-data", such as reference to issue, workflow actions and time logged, belong in the commit footer (rather than the title).

**EDIT**: Further research leads me to believe your default commit format is somewhat similar to JIRA's own [Smart Commits](https://confluence.atlassian.com/display/BITBUCKET/Processing+JIRA+issues+with+commit+messages#ProcessingJIRAissueswithcommitmessages-Understandtherequirementsofsmartcommits) used in their [DVCS connector plugin](https://marketplace.atlassian.com/plugins/com.atlassian.jira.plugins.jira-bitbucket-connector-plugin). If we want to take it a step forward, I would use their format exclusively while allowing any line of the commit message (and not just the beginning) to be parsed for matching tokens.

This should approach will allow me to use a commit message like this:

``` git
fix(app): correct out of scope variable

my long description

JRA-123 #time 2d 5h #comment Task completed ahead of schedule #resolve
```

**Comment by carlosabalde** [comment]
At the moment you should be able to use the `#time` and `#action` directives at the bottom of commit messages. Personally, I like the idea of adding all that info in there.

Additionally, I think it would be nice being able to reference issue identifiers anywhere inside commit messages using a new directive (e.g. `#refs`). This is a simple enhancement that would help to keep commit messages clean. This way a commit messages will look like:

```
Short commit description

Commit details

#refs ALLGEN-10 ALLGEN-42 #time 2d 5h #action close issue, fixed
```

Of course, you'll still be able to include the issue identifiers at the beginning of the message without the `#refs` directive. If you agree this is useful I'll add these ideas to the TODO list.

**Comment by asfaltboy** [comment]
Yes please :+1:  this would be good workaround

**Comment by asfaltboy** [comment]
@carlosabalde : by the way, can you provide an ETA for this enhancement? We have purchased the addon but are currently unable to use it (at least not without changing our commit message consensus)...

**Comment by carlosabalde** [comment]
It's now available in the JIRA marketplace (v1.1.5) :tada:

**Comment by asfaltboy** [comment]
Very nice, tested and works well! Thanks for your help
